diff --git a/SConstruct b/SConstruct
index 1e175c3..08de684 100755
--- a/SConstruct
+++ b/SConstruct
@@ -282,6 +282,7 @@ if env["prereqs"]:
         if conf.CheckCHeader("sys/sendfile.h", "<>"):
             conf.CheckFunc("sendfile")
     conf.CheckLib("m")
+    conf.CheckLib("pthread")
     conf.CheckFunc("round")
 
     have_server_prereqs = \
diff --git a/build.sh b/build.sh
new file mode 100755
index 0000000..5027b86
--- /dev/null
+++ b/build.sh
@@ -0,0 +1,14 @@
+#!/bin/bash
+set -e -x
+
+ROOT=${NACL_TOOLCHAIN_ROOT:-$HOME/root/nacl-sdk}
+PATH=$ROOT/bin:$ROOT/nacl64/usr/bin:$PATH
+
+# export SDL_LIBS="-L$ROOT/nacl64/usr/lib -lSDL -lppapi_cpp -lpthread -lm"
+PKG_CONFIG_PATH=$ROOT/nacl64/usr/lib/pkgconfig scons --config=force host=nacl64 prefix=$ROOT/nacl64/usr boostdir=$ROOT/nacl64/usr/include/boost boostlibdir=$ROOT/nacl64/usr/lib sdldir=$ROOT/nacl64/usr
+
+# CC=nacl64-gcc AR=nacl64-ar RANLIB=nacl64-ranlib PKG_CONFIG_LIBDIR=$ROOT/nacl64/usr/lib/pkgconfig \
+#   ./configure --host=nacl64 --prefix=$ROOT/nacl64/usr --without-x
+# make
+# make install
+
diff --git a/scons/config_check_utils.py b/scons/config_check_utils.py
index b566791..5484ad5 100644
--- a/scons/config_check_utils.py
+++ b/scons/config_check_utils.py
@@ -14,5 +14,5 @@ def restore_env(env, backup):
 
 def find_include(prefixes, include_file, include_subdir, default_prefixes = True):
     if default_prefixes:
-        prefixes = ["/usr", "/usr/local", "/sw", "/sw/local"] + prefixes
+        prefixes = prefixes + ["/usr", "/usr/local", "/sw", "/sw/local"]
     return [(prefix, include) for prefix in prefixes for include in glob(join(prefix, "include", include_subdir, include_file))]
